pkg/gameservers/controller.go

```diff
@@ -761,27 +761,42 @@ func (c *Controller) syncGameServerRequestReadyState(gs *agonesv1.GameServer) (*
 	c.loggerForGameServer(gs).Info("Syncing RequestReady State")
 
 	gsCopy := gs.DeepCopy()
+	pod, err := c.gameServerPod(gs)
+	// The most likely reason this fails is if the master has gone down, as it's very unlikely
+	// the Pod got deleted between sending `RequestReady` and now,
+	// so go into backoff until whatever the error is resolves itself.
+	if err != nil {
+		return nil, err
+	}
+
+	// track the ready gameserver container, so we can determine that after this point, we should move to Unhealthy if
+	// there is a container crash/restart after we move to Ready
+	for _, cs := range pod.Status.ContainerStatuses {
+		if cs.Name == gs.Spec.Container {
+			if _, ok := pod.ObjectMeta.Annotations[agonesv1.GameServerReadyContainerIDAnnotation]; !ok {
+				pod.ObjectMeta.Annotations[agonesv1.GameServerReadyContainerIDAnnotation] = cs.ContainerID
+				if _, err = c.podGetter.Pods(gs.ObjectMeta.Namespace).Update(pod); err != nil {
+					return nil, errors.Wrapf(err, "error adding ready containerid to pod %s", pod.ObjectMeta.Name)
```

Comment:
qq: Why apply the annotation to the pod instead of to the gameserver? If we applied it to the game server then you could combine this with the update call below, which would make the state transition to ready (and this container identification) happen in a single API call. Right now there is a slight chance that this succeeds and the following call fails (or never happens due to a crash) and then we've marked the container id but never moved the game server to ready.

Reply:
That's a really good question - so in [health.go](https://github.com/googleforgames/agones/pull/1069/files#diff-c3108a3059e5decdd651ad0d0b47479eR107) we've only ever looked up the Pod.

In _theory_ we could also lookup the GameServer and check it's status. Maybe that's actually a better option. Since it's likely in the Informer cache anyway. There are only a few states (maybe just one? Scheduled) that come before being Ready, so there aren't a lot of edge cases to check -- and then we don't need to even check containerids.

The more I think about it, this seems like it might be a better approach. How does that sound to you?

Reply:
Digging into this, this would need to change the strategy on the controller - as there could be an error getting the GameServer, so we may have to push this into a queue (whereas we don't have to do it right now)

Context:
https://github.com/googleforgames/agones/blob/2c8cc9bb18bce4b8bafa3d61246b0980e058c9e6/pkg/gameservers/health.go#L87

So there are pros and cons both ways.